GitExtUtils/FileLogging/Logger.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GitExtUtils.FileLogging
+{
+    /// <summary>
+    /// For unit testing
+    /// </summary>
+    public class LoggerStub : ILogger
+    {
+        public List<LogEntry> LogEntries = new List<LogEntry>();
+
+        public void WaitShutdown()
+        {
+        }
+
+        public void Log(LogSeverity severity, string message, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(severity, message, extra));
+        }
+
+        public void Info(string message, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(LogSeverity.Info, message, extra));
+        }
+
+        public void Error(string message, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(LogSeverity.Error, message, extra));
+        }
+
+        public void Error(string message, Exception e, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(LogSeverity.Error, message, e, extra));
+        }
+
+        public void LogGitCommandExecution(string message, string fileName, string arguments, string workingDir, int? processId,
+            bool isOnMainThread, TimeSpan? duration, int? exitCode, StackTrace callStack)
+        {
+            LogEntries.Add(LogEntryCreator.Create(message, fileName, arguments, workingDir, processId, isOnMainThread, duration,
+                exitCode, callStack));
+        }
+    }
+
+    public class Logger : ILogger
+    {
+        private readonly LogWriter _writer;
+
+        private readonly Task _writerTask;
+
+        public const string MetaTypeKeyName = "_MetaType";
+
+        private static ILogger _instance = new LoggerStub();
+
+        public Logger(
+            string gitExtensionsPath,
+            ILogFormatter formatter,
+            TimeSpan flushInterval,
+            TimeSpan maxLogAge,
+            CancellationToken cancellationToken)
+        {
+            if (!Directory.Exists(gitExtensionsPath))
+            {
+                throw new ArgumentException($"Non-existing folder '{gitExtensionsPath}'", nameof(gitExtensionsPath));
+            }
+
+            var logPath = Path.Combine(gitExtensionsPath, "logs");
+            var creator = new LogPathGenerator(logPath, "gitextensions");
+            _writer = new LogWriter(creator, formatter, flushInterval);
+
+            Task.Run(new LogFileCleaner(this, creator, maxLogAge).Run);
+
+            _writerTask = Task.Run(() => _writer.Run(cancellationToken));
+
+            _instance = this;
```

Comment:
If we instantiate this class multiple times, we will be changing the static instance.... This doesn't makes sense

Reply:
But we are only instantiating it once - from the composition root of the application. Why would you make multiple instances? Are you thinking from unit tests? We can salvage with a `lock` to ensure it is only ever set once.